File: common/src/main/scripts/lib/lua/oc2/blob.lua
local fcntl = require("posix.fcntl")
local unistd = require("posix.unistd")
local poll = require("posix.poll")
local json_null = require("cjson").null
local Channel = require("oc2.channel")

local blob = {}

local chunkSize = 32 * 1024
local readTimeout = 5000
local maxDepth = 32

blob.key = "$blob"
blob.maxOutbound = 512 * 1024
blob.maxInbound = blob.maxOutbound
blob.chunkSize = chunkSize
blob.readTimeout = readTimeout
blob.maxDepth = maxDepth
blob.marker = {}

function blob.wrap(data)
  if type(data) ~= "string" then
    error("a binary payload must be a string, got " .. type(data), 2)
  end
  return setmetatable({ data = data }, blob.marker)
end

function blob.checksum(data)
  local sum = 0
  local length = #data
  local i = 1
  while i <= length do
    local word
    if i + 3 <= length then
      word = string.unpack("<I4", data, i)
    else
      word = string.unpack("<I4", data:sub(i) .. string.rep("\0", 4 - (length - i + 1)))
    end
    sum = (((sum << 1) | (sum >> 31)) + word) & 0xFFFFFFFF
    i = i + 4
  end
  if sum >= 0x80000000 then
    sum = sum - 0x100000000
  end
  return sum
end

function blob.extract(...)
  local packed = table.pack(...)
  local parameters = {}
  local payload
  for i = 1, packed.n do
    local value = packed[i]
    if getmetatable(value) == blob.marker then
      if payload then
        error("a call may carry at most one binary payload", 2)
      end
      payload = value.data
      parameters[i] = { [blob.key] = true }
    elseif value == nil then
      parameters[i] = json_null
    else
      parameters[i] = value
    end
  end
  return parameters, payload
end

function blob.substitute(value, payload, depth)
  if type(value) ~= "table" then
    return value
  end

  depth = depth or 0
  if depth > maxDepth then
    error("host reply is nested too deeply", 0)
  end

  if value[blob.key] then
    return payload
  end

  for key, item in pairs(value) do
    value[key] = blob.substitute(item, payload, depth + 1)
  end
  return value
end

local Payload = {}
Payload.__index = Payload

function blob.fromFd(fd)
  return setmetatable({ fd = fd }, Payload)
end

function blob.open(path)
  local fd, status = fcntl.open(path, fcntl.O_RDWR | fcntl.O_CLOEXEC | fcntl.O_NONBLOCK)
  if not fd then
    return nil, status
  end
  return blob.fromFd(fd)
end

function Payload:close()
  if self.fd then
    unistd.close(self.fd)
    self.fd = nil -- closing twice would close whatever reused the number
  end
end

function Payload:reset()
  repeat
    local ready = poll.rpoll(self.fd, 0)
    if ready == 1 then
      local data = unistd.read(self.fd, chunkSize)
      if not data or #data == 0 then
        break
      end
    end
  until ready ~= 1
end

function Payload:write(data)
  if #data > blob.maxOutbound then
    error(string.format("binary payload of %d bytes exceeds the host limit of %d bytes",
                        #data, blob.maxOutbound), 2)
  end

  Channel.writeAll(self.fd, data)
end

function Payload:read(length, timeout)
  return Channel.readExactly(self.fd, length, timeout or readTimeout)
end

function blob.resolve(channel, result)
  local reference = result.blob
  if not reference then
    return result.data
  end

  if reference.length < 0 or reference.length > blob.maxInbound then
    error("host announced an implausible payload size: " .. tostring(reference.length), 0)
  end

  local data, reason = channel:read(reference.length)
  if not data then
    channel:reset()
    error("could not read binary payload: " .. tostring(reason), 0)
  end
  if blob.checksum(data) ~= reference.checksum then
    channel:reset()
    error("binary payload failed its checksum; the data channel is corrupt or out of sync", 0)
  end

  return blob.substitute(result.data, data)
end

return blob

File: common/src/main/scripts/lib/micropython/oc2/blob.py
import io
import select

from oc2 import ports
from oc2.channel import write_all

KEY = "$blob"
MAX_OUTBOUND_SIZE = 512 * 1024
CHUNK_SIZE = 4096
READ_TIMEOUT_MS = 5000
MAX_INBOUND_SIZE = MAX_OUTBOUND_SIZE


class Blob:
    def __init__(self, data):
        self.data = data.encode() if isinstance(data, str) else data


def checksum(data):
    total = 0
    length = len(data)

    whole = length - (length % 4)
    for offset in range(0, whole, 4):
        word = (data[offset] | (data[offset + 1] << 8)
                | (data[offset + 2] << 16) | (data[offset + 3] << 24))
        total = ((total << 1) | (total >> 31)) & 0xFFFFFFFF
        total = (total + word) & 0xFFFFFFFF

    if whole < length:  # zero-padded tail, once
        word = 0
        for i in range(whole, length):
            word |= data[i] << ((i - whole) * 8)
        total = ((total << 1) | (total >> 31)) & 0xFFFFFFFF
        total = (total + word) & 0xFFFFFFFF

    return total - 0x100000000 if total >= 0x80000000 else total


def extract(args):
    payload = None
    parameters = []
    for value in args:
        if isinstance(value, Blob):
            if payload is not None:
                raise Exception("a call may carry at most one binary payload")
            payload = value.data
            parameters.append({KEY: True})
        elif isinstance(value, (bytes, bytearray)):
            # json.dumps would quietly encode these as a string. Binary has to be declared,
            # so it travels on the payload channel and the host sees a byte array.
            raise Exception("wrap binary parameters in bus.blob(...)")
        else:
            parameters.append(value)
    return parameters, payload


def substitute(value, payload, depth=0):
    if depth > 32:
        raise Exception("host reply is nested too deeply")

    if isinstance(value, dict):
        if value.get(KEY):
            return payload
        for key in value:
            value[key] = substitute(value[key], payload, depth + 1)
    elif isinstance(value, list):
        for i in range(len(value)):
            value[i] = substitute(value[i], payload, depth + 1)
    return value


class PayloadChannel:
    def __init__(self, path):
        self.file = io.open(path, "+b")
        ports.set_nonblocking(self.file.fileno())
        self.poll = select.poll()
        self.poll.register(self.file.fileno(), select.POLLIN)
        self.write_poll = select.poll()
        self.write_poll.register(self.file.fileno(), select.POLLOUT)

    def close(self):
        if self.file:
            self.poll.unregister(self.file.fileno())
            self.file.close()
            self.file = None  # closing twice would close whatever reused the number

    def reset(self):
        while self.file.read(CHUNK_SIZE):
            pass

    def write(self, data):
        if len(data) > MAX_OUTBOUND_SIZE:
            raise Exception("binary payload of %d bytes exceeds the host limit of %d bytes"
                            % (len(data), MAX_OUTBOUND_SIZE))

        write_all(self.file, self.write_poll, data)

    def read(self, length):
        parts = bytearray()
        while len(parts) < length:
            if not self.poll.poll(READ_TIMEOUT_MS):
                raise Exception("timed out waiting for the rest of the payload")
            chunk = self.file.read(min(length - len(parts), CHUNK_SIZE))
            if chunk is None:
                continue  # readable, but nothing yet
            if not chunk:
                raise Exception("end of file on the binary channel")
            parts.extend(chunk)
        return bytes(parts)


def resolve(channel, message):
    reference = message.get("blob")
    if reference is None:
        return message.get("data")

    if channel is None:
        raise Exception("host sent a binary payload but no data channel was found")

    length = reference["length"]
    if length < 0 or length > MAX_INBOUND_SIZE:
        raise Exception("host announced an implausible payload size: %s" % length)

    data = channel.read(length)
    if checksum(data) != reference.get("checksum"):
        channel.reset()
        raise Exception("binary payload failed its checksum; "
                        "the data channel is corrupt or out of sync")

    return substitute(message.get("data"), data)
